library/core/src/slice/mod.rs

```diff
@@ -2927,15 +2927,7 @@ impl<T> [T] {
     where
         T: Clone,
     {
-        assert!(self.len() == src.len(), "destination and source slices have different lengths");
-        // NOTE: We need to explicitly slice them to the same length
-        // for bounds checking to be elided, and the optimizer will
-        // generate memcpy for simple cases (for example T = u8).
```

Comment:
This doesn't (always) happen. While profiling a testsuite I saw a very dumb copy loop showing up in gimli. I have [submitted a fix](https://github.com/gimli-rs/gimli/pull/549) for gimli specifically but this should apply more broadly and help generic code.